### PR TITLE
Support CodeContracts

### LINKED ISSUE #389: Support CodeContracts
Currently there is no support for CodeContracts.
http://research.microsoft.com/en-us/projects/contracts/

### PR DESCRIPTION
**Pull request by Mudhoax**
Fixes #389

Adds CodeContract support; should be noted that only some run-time features are supported as to avoid any rewriter implementation. Although as far as I can tell this allows for full static analysis support.

### CODE REVIEW

**Comment by Mudhoax** [comment]
Spotted a bug in this at present.
When a project references another project which is using this CodeContracts implementation I am getting the following error:

    The serializable type System.Diagnostics.Contracts.__ContractsRuntime.ContractException must inherit from another serializable type, System.Object or System.Record.

I will need to investigate this one further.

File: Runtime/CoreLib/Collections/Generic/KeyValuePair.cs [hunk 1]
```diff
@@ -9,7 +9,7 @@ namespace System.Collections.Generic {
 	[IncludeGenericArguments(false)]
 	[Serializable]
 	[Imported]
-	public sealed class KeyValuePair<TKey, TValue> {
+	public struct KeyValuePair<TKey, TValue> {
 		public KeyValuePair(TKey key, TValue value) {
 		}
```
Comment (erik-kallen) [inline Runtime/CoreLib/Collections/Generic/KeyValuePair.cs]: Why this change? I don't think it hurts but why is it necessary?

File: Runtime/CoreLib/CoreLib.csproj [hunk 2]
```diff
@@ -264,5 +279,8 @@
       <Link>CoreLib.Plugin.dll</Link>
     </EmbeddedResource>
   </ItemGroup>
+  <ItemGroup>
+    <Reference Include="System" />
+  </ItemGroup>
   <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
```
Comment (erik-kallen) [inline Runtime/CoreLib/CoreLib.csproj]: This reference is bad.

Comment (Mudhoax) [inline Runtime/CoreLib/CoreLib.csproj]: Yes; I will remove this. My bad.

File: Runtime/CoreLib/Runtime.cs [hunk 3]
```diff
@@ -25,7 +25,8 @@ public abstract class MarshalByRefObject {
 
 	[EditorBrowsable(EditorBrowsableState.Never)]
 	[Imported]
-	[IgnoreNamespace]
+    [IgnoreNamespace]
+    [Serializable]
 	[ScriptName("Object")]
```
Comment (erik-kallen) [inline Runtime/CoreLib/Runtime.cs]: Why is this necessary?

File: Runtime/CoreLib/Collections/Generic/KeyValuePair.cs (hunk 1 above)
Comment (Mudhoax) [inline Runtime/CoreLib/Collections/Generic/KeyValuePair.cs]: The CodeContracts extension attempts to convert this to a struct and fails.

**Comment by erik-kallen** [comment]
The problem you are experiencing is most likely due to the reference to System that was accidentally added.

I don't know a lot about code contracts, but do they really belong in mscorlib? or should we create a separate project/dll for them?

**Comment by Mudhoax** [comment]
CodeContracts were moved into the mscorlib. I figured it would be worth a shot at adding it into the core runtime here. However if you would prefer this to be a plugin then please let me know.

File: Runtime/CoreLib/Runtime.cs (hunk 3 above)
Comment (Mudhoax) [inline Runtime/CoreLib/Runtime.cs]: With the change of KVP to a struct this is required to still allow KVP to be serializable.

File: Runtime/CoreLib/Collections/Generic/KeyValuePair.cs (hunk 1 above)
Comment (erik-kallen) [inline Runtime/CoreLib/Collections/Generic/KeyValuePair.cs]: Where and how? Not saying you're wrong, just wondering. Are you talking about a Visual Studio extension or an extension method?

**Comment by erik-kallen** [comment]
Hmm, I'm not sure.

If they are in mscorlib, they can go in mscorlib here as well. But perhaps the runtime support can be in another JS file? What do you think?

**Comment by Mudhoax** [comment]
The runtime support is configurable via the CodeContracts extension; and can be turned off completely.
Our usages are purely used within debug and never enter our release builds.

If you could walk me through the process of moving the runtime support to a separate js file and would still prefer that approach I would be happy to do so.

**Comment by Mudhoax** [comment]
Removing the System reference still doesn't resolve the previous issue. Will look into this further tomorrow.

File: Runtime/CoreLib/Collections/Generic/KeyValuePair.cs (hunk 1 above)
Comment (Mudhoax) [inline Runtime/CoreLib/Collections/Generic/KeyValuePair.cs]: A visual studio extention. https://visualstudiogallery.msdn.microsoft.com/1ec7db13-3363-46c9-851f-1ce455f66970

Comment (Mudhoax) [inline Runtime/CoreLib/Collections/Generic/KeyValuePair.cs]: ....
    GenericKeyValuePair = (Struct)GetGenericRuntimeTypeNodeFor("System.Collections.Generic", "KeyValuePair", 2, ElementType.ValueType);
    ....

https://github.com/Microsoft/CodeContracts/blob/master/System.Compiler/SystemTypes.cs

**Comment by Mudhoax** [comment]
So further investigation to the issue I've uncovered that it's not really a problem but something that needs documenting somewhere. Basically its the CodeContracts runtime injection that appears to be incompatible; this isn't really a problem since all I was using it for was it define the CONTRACTS_FULL symbol in DEBUG. 
image
Simply ensuring that Runtime Checking is disabled ensures that no code will be injected and solves the issue.
If however anybody still wants the runtime code to be injected we can simply define the compilation symbol as part of out build.
image
It should be noted that we have only ever encountered this as a issue when attempting to build a project that is referenced to a project using CodeContracts. This should not affect anybody that does not have the CodeContracts extension installed.

Would you classify this as a acceptable workaround?
In addition let me know if you need anything further.

**Comment by erik-kallen** [comment]
We could probably just change the MetadataImporter to not process types in the System.Diagnostics.Contracts.__ContractsRuntime assembly.

**Comment by Mudhoax** [comment]
Thanks for your response, could you provide some more information as to how I would go about implementing something along those lines? Sorry I am still fairly unfamiliar to all the internals of the transpiler.

**Comment by erik-kallen** [comment]
Inside the ```MetadataImporter``` class there is a method called ProcessType. Add an if statement to the top of that method (```if (type.Namespace == "System.Diagnostics.Contracts.__ContractsRuntime") return TypeScriptSemantics.NotUsableFromScript();```)

**Comment by Mudhoax** [comment]
Thanks for that. Although it doesn't seem to make a difference. I tried:

    if (typeDefinition.Namespace == "System.Diagnostics.Contracts.__ContractsRuntime")
    {
        _typeSemantics[typeDefinition] = new TypeSemantics(TypeScriptSemantics.NotUsableFromScript(), false, false, false);
        return;
    }

I can't seem to attach to any of the build processes to get any further details unfortunately. Will look into this a bit more if I get some time later.

**Comment by Mudhoax** [comment]
Latest commit appears to fix the problem. Funnily enough the assertion is correct when checking the types FullName as apposed to its namespace.

**Comment by erik-kallen** [comment]
Will merge this manually into the 3.0 branch
